validator_client/src/doppelganger_service.rs

```diff
@@ -67,95 +321,1026 @@ impl<T: 'static + SlotClock, E: EthSpec> DoppelgangerService<T, E> {
+                    validator on the network and is shutting down to prevent potential slashable \
+                    offences. Ensure that you are not running a duplicate or overlapping \
+                    validator client",
+                "doppelganger_indices" => ?violators
+            )
+        }
+
+        // The concept of "epoch satisfaction" is that for some epoch `e` we are *satisfied* that
+        // we've waited long enough such that we don't expect to see any more consensus messages
+        // for that epoch.
+        //
+        // As it stands now, we consider epoch `e` to be satisfied once we're in the last slot of
+        // epoch `e + 1`.
+        //
+        // The reasoning for this choice of satisfaction slot is that by this point we've
+        // *probably* seen all the blocks that are permitted to contain attestations from epoch `e`.
+        let previous_epoch_satisfaction_slot = previous_epoch
+            .saturating_add(1_u64)
+            .end_slot(E::slots_per_epoch());
+        let previous_epoch_is_satisfied = request_slot >= previous_epoch_satisfaction_slot;
+
+        // Iterate through all the previous epoch responses, updating `self.doppelganger_states`.
+        //
+        // Do not bother iterating through the current epoch responses since they've already been
+        // checked for violators and they don't result in updating the state.
+        for response in &previous_epoch_responses {
+            // Sanity check response from the server.
+            //
+            // Abort the entire routine if the server starts returning junk.
+            if response.epoch != previous_epoch {
+                return Err(format!(
+                    "beacon node returned epoch {}, expecting {}",
+                    response.epoch, previous_epoch
+                ));
+            }
+
+            // Resolve the index from the server response back to a public key.
+            let pubkey = indices_map
+                .get(&response.index)
+                // Abort the routine if inconsistency is detected.
+                .ok_or_else(|| {
+                    format!(
+                        "inconsistent indices map for validator index {}",
+                        response.index
+                    )
+                })?;
+
+            // Hold the lock on `self` for the rest of this function.
+            //
+            // !! IMPORTANT !!
+            //
+            // There is a write-lock being held, avoid interacting with locks until it is dropped.
+            let mut doppelganger_states = self.doppelganger_states.write();
+            let doppelganger_state = doppelganger_states
+                .get_mut(&pubkey)
+                // Abort the routine if inconsistency is detected.
+                .ok_or_else(|| format!("inconsistent states for validator pubkey {}", pubkey))?;
+
+            // If a single doppelganger is detected, enable doppelganger checks on all
```

Comment:
I think if we're able to do this on a validator by validator basis it'd be better

Reply:
Since we're going to shutdown the VC once we detect any doppelganger, aren't we effectively locked into the "all or none" behaviour?

The reason I'm jamming all of them here is so it's more firmly "all or none" not a "some until the VC finishes whatever it's doing".

Reply:
My thinking here was related to my thinking below about maybe trying to recover if we can't shutdown.. but I am now thinking that is too risky.  I am understanding the reasoning for this logic a bit more, hadn't considered the impact of sigp#2316